promql/prettier/node.go

```diff
@@ -0,0 +1,182 @@
+			nodeMatch = true
+			stack = append(stack, n)
+			return p.nodeHistory(n.Expr, posRange, stack)
+		}
+	case *parser.VectorSelector:
+		if n.PositionRange().Start <= posRange.Start && n.PositionRange().End >= posRange.End {
+			nodeMatch = true
+			stack = append(stack, n)
+		}
+	case *parser.BinaryExpr:
+		if n.PositionRange().Start <= posRange.Start && n.PositionRange().End >= posRange.End {
+			nodeMatch = true
+			stack = append(stack, n)
+			stmp, node, found := p.nodeHistory(n.LHS, posRange, stack)
+			if found {
+				return stmp, node, found
+			}
+			stmp, node, found = p.nodeHistory(n.RHS, posRange, stack)
+			if found {
+				return stmp, node, found
+			}
+			// Since the item exists in both the child. This means that it is in binary expr range,
+			// but not satisfied by a single child. This is possible only for Op and grouping
+			// modifiers.
+			if n.PositionRange().Start <= posRange.Start && n.PositionRange().End >= posRange.End {
+				return stack, head, true
+			}
+		}
+	case *parser.AggregateExpr:
+
+		if n.PositionRange().Start <= posRange.Start && n.PositionRange().End >= posRange.End {
+			nodeMatch = true
+			stack = append(stack, n)
+			stmp, _head, found := p.nodeHistory(n.Expr, posRange, stack)
+			if found {
+				return stmp, _head, true
+			}
+			if n.Param != nil {
+				if stmp, _head, found := p.nodeHistory(n.Param, posRange, stack); found {
+					return stmp, _head, true
+				}
+			}
+		}
+	case *parser.Call:
+		if n.PositionRange().Start <= posRange.Start && n.PositionRange().End >= posRange.End {
+			nodeMatch = true
+			stack = append(stack, n)
+			for _, exprs := range n.Args {
+				if exprs.PositionRange().Start <= posRange.Start && exprs.PositionRange().End >= posRange.End {
+					stmp, _head, found := p.nodeHistory(exprs, posRange, stack)
+					if found {
+						return stmp, _head, true
+					}
+				}
+			}
+		}
+	case *parser.MatrixSelector:
+		if n.VectorSelector.PositionRange().Start <= posRange.Start && n.VectorSelector.PositionRange().End >= posRange.End {
+			stack = append(stack, n)
+			nodeMatch = true
+			p.nodeHistory(n.VectorSelector, posRange, stack)
+		}
+	case *parser.UnaryExpr:
+		stack = append(stack, n)
+		if n.Expr.PositionRange().Start <= posRange.Start && n.Expr.PositionRange().End >= posRange.End {
+			nodeMatch = true
+			p.nodeHistory(n.Expr, posRange, stack)
+		}
+	case *parser.SubqueryExpr:
+		if n.Expr.PositionRange().Start <= posRange.Start && n.Expr.PositionRange().End >= posRange.End {
+			nodeMatch = true
+			stack = append(stack, n)
+			p.nodeHistory(n.Expr, posRange, stack)
+		}
+	case *parser.NumberLiteral, *parser.StringLiteral:
+		nodeMatch = true
+		stack = append(stack, n)
+	}
+	return stack, head, nodeMatch
+}
+
+// reduceContinuous reduces from end, the continuous
+// occurrence of a type to its single representation.
+func reduceContinuous(history []parser.Node, typString string) []parser.Node {
```

Comment:
This is only used for binary expressions, right?

So you could avoid the reflect package by specializing this function.

Also, this would make it easier to implement precedence based reduction

Reply:
Implemented this, hence resolving.